src/routers/account.js

```diff
@@ -494,4 +499,150 @@ router.get('/miieditor', async (request, response) => {
+	});
+
+	if (apiResponse.statusCode !== 200) {
+		// Assume expired, refresh and retry request
+		apiResponse = await util.apiPostGetRequest('/v1/login', {}, {
+			refresh_token: request.cookies.refresh_token,
+			grant_type: 'refresh_token'
+		});
+
+		if (apiResponse.statusCode !== 200) {
+			return response.redirect('/account/login');
+		}
+
+		const tokens = apiResponse.body;
+
+		response.cookie('refresh_token', tokens.refresh_token, { domain: '.pretendo.network' });
+		response.cookie('access_token', tokens.access_token, { domain: '.pretendo.network' });
+		response.cookie('token_type', tokens.token_type, { domain: '.pretendo.network' });
+
+		apiResponse = await util.apiGetRequest('/v1/user', {
+			'Authorization': `${tokens.token_type} ${tokens.access_token}`
+		});
+	}
+
+	// If still failed, something went horribly wrong
+	if (apiResponse.statusCode !== 200) {
+		return response.redirect('/account/login');
+	}
+
+	// Set user account info to render data
+	const account = apiResponse.body;
+	const pid = account.pid;
+
+	let customer;
+	const { data: searchResults } = await await stripe.customers.search({
+		query: `metadata['pnid_pid']:'${pid}'`
+	});
+
+	if (searchResults.length !== 0) {
+		customer = searchResults[0];
+	} else {
+		customer = await stripe.customers.create({
+			metadata: {
+				pnid_pid: pid
+			}
+		});
+	}
+
+	const priceId = request.params.priceId;
+
+	const session = await stripe.checkout.sessions.create({
+		line_items: [
+			{
+				price: priceId,
+				quantity: 1,
+			},
+		],
+		customer: customer.id,
+		mode: 'subscription',
+		success_url: `${config.http.base_url}/account?upgrade_success=true`,
+		cancel_url: `${config.http.base_url}/account?upgrade_success=false`
+	});
+
+	response.redirect(303, session.url);
+});
+
+router.post('/stripe-wh', express.raw({ type: 'application/json' }), async (request, response) => {
+	const stripeSignature = request.headers['stripe-signature'];
+	let event;
+
+	try {
+		event = stripe.webhooks.constructEvent(request.body, stripeSignature, config.stripe.webhook_secret);
+	} catch (err) {
+		return response.status(400).send(`Webhook Error: ${err.message}`);
+	}
+
+	if (event.type === 'customer.subscription.updated' || event.type === 'customer.subscription.deleted') {
+		const subscription = event.data.object;
+		const product = await stripe.products.retrieve(subscription.plan.product);
+		const customer = await stripe.customers.retrieve(subscription.customer);
+		const pid = Number(customer.metadata.pnid_pid);
+		const pnid = await database.PNID.findOne({ pid });
+
+		if (product.metadata.beta === 'true') {
+			switch (subscription.status) {
+				case 'active':
+					pnid.access_level = 1;
+					break;
+
+				case 'canceled': // Subscription was cancled
+				case 'unpaid': // User missed too many payments
+					pnid.access_level = 0;
+					break;
+			
+				default:
+					break;
+			}
+
+			await pnid.save();
```

Comment:
Events can be received out of order.
Use a timestamp to "ignore" older events.

https://stripe.com/docs/webhooks/best-practices#event-ordering

Reply:
Can you clarify why this is an issue here? Is there a concern with, say, an event with the `active` status coming in after one with `unpaid` for example?

I’m just trying to fully understand the issue before implementing a fix for it

Reply:
Imagine someone's payment failing, then doing a retry a few minutes after. (Quite common in payment, called soft fails)

That would send two events: unpaid and active. With a few minutes apart.

If the first one fails for any reason (outage, restart, deploy, crash, etc) and the second one succeeds. It will all be fine, the problem starts when stripe does it's retry logic an hour later. It will then send the `unpaid` status and will mark the account as not having paid, even tho they did pay successfully.

By storing the last sync time in the db (gotten from event), you can ignore older events that aren't relevant anymore.

Reply:
I wouldn't put it on the customer meta data for two reasons:

1.its a seperate call, then you have two save calls. what is your backup plan if one of them fails?
if either fails, it will result in the account db being wrong in the end (either syncing wrong or being overwritten by an old event)

2. its not related to the customer, its related to the account, not the customer. those are differen entities, even if they refer to the same person.
 
---

Possible solutions:
1. go full microservice mode, have a seperate service with a seperate db for all payment and access level storage.
when any service needs to know the data, they query the microservice.

2. store the timestamp on the account db (easiest, i recommend this one)
sure its not technically related to the account the PNID. but you gotta have some place to store it.

Reply:
> its a seperate call, then you have two save calls. what is your backup plan if one of them fails?
if either fails, it will result in the account db being wrong in the end (either syncing wrong or being overwritten by an old event)

What do you mean separate calls? I'm not pulling the customer data twice, I already pull it once? And I'm not sure what you mean by the `it will result in the account db being wrong in the end` since the entire point of storing it on the customer is to _not_ have it in the account DB? There's nothing there to "be wrong"?

> its not related to the customer, its related to the account, not the customer. those are differen entities, even if they refer to the same person.

I would say it _is_ related to the customer. It's the Stripe customer which is triggering the webhook, not the PNID? So would it not make sense to just store the data relating to webhook the customer triggered on the webhook itself?

> go full microservice mode, have a seperate service with a seperate db for all payment and access level storage.
when any service needs to know the data, they query the microservice.

That is very. Over-engineering imo?

> store the timestamp on the account db (easiest, i recommend this one)
sure its not technically related to the account the PNID. but you gotta have some place to store it.

The fact that it's not related to the PNID is exactly why I don't want to store it there? It's related to the customer

Reply:
> What do you mean separate calls?

you have to update the customer meta data with an api call. and you have to update the access level on the account db document. those can both fail.

and `There's nothing there to "be wrong"?`, what can be wrong is the wrong access level on the PNID.
if the timestamp on customer meta-data fails to update then an older event on a retry can insert wrong access level.
if the access level fails to save but the timestamp does update, they will never obtain that acesss level back until next payment flow.

Databases like postgres and mongo have to follow ACID principles. the A of ACID stands for atomicity. meaning that they can save the entire statement or nothing at all. it will not save half the data. which is exactly what you need right now.
you dont have a garuantee for it be atomic if you have two seperate calls between your pieces of data. so you have to store it in one DB.

> That is very. Over-engineering imo?
depends on your architecture. for pretendo, yes it is.

> The fact that it's not related to the PNID is exactly why I don't want to store it there? It's related to the customer
what other option do you have?